### PR TITLE
`if $c:expr { Some($r:expr) } else { None }` =>> `$c.then(|| $r)`

### PR DESCRIPTION
**Pull request by WaffleLapkin**
TL;DR: turn
```rust
if condition {
    Some(expression)
} else {
   None
}
```
into
```rust
condition.then(|| expression)
```

This is part one of many of refactoring compiler to use `.then` where it makes code cleaner. This PR does only changes that are simple & clear wins (in my opinion anyway).

### CODE REVIEW

File: compiler/rustc_ast_passes/src/ast_validation.rs [hunk 1]
```diff
@@ -271,7 +271,7 @@ impl<'a> AstValidator<'a> {
 
         self.session.emit_err(InvalidVisibility {
             span: vis.span,
-            implied: if vis.kind.is_pub() { Some(vis.span) } else { None },
+            implied: vis.kind.is_pub().then(|| vis.span),
```
Comment (oli-obk) [inline compiler/rustc_ast_passes/src/ast_validation.rs:274]: These cases may be simpler as `then_some` (even if longer 🙃 ). I don't have a strong opinion on `then_some` usage, beyond it generating less code

**Review by oli-obk — APPROVED** [review]

**Comment by slanterns** [comment]
r+ when doing `try` build will cause the pr merged without going through testing🙃 And it blocks #101841.

**Comment by matthiaskrgr** [comment]
Perhaps rustbot could add the `Blocked` label when posting "Awaiting bors try build completion." comment and remove it again when posting the final  perf summary

**Pull request merged**